program/src/processor.rs

```diff
@@ -407,49 +413,27 @@ fn process_harvest_rewards(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
         bytemuck::try_from_bytes_mut::<HolderRewards>(&mut holder_rewards_data)
             .map_err(|_| ProgramError::InvalidAccountData)?;
 
-    // Update the holder rewards state with the new "last seen" total rewards.
-    let last_seen_total_rewards = std::mem::replace(
-        &mut holder_rewards_state.last_seen_total_rewards,
-        current_total_rewards,
-    );
-
-    // Calculate unharvested rewards for the token account.
-    // Since the holder rewards account may already have unharvested rewards,
-    // calculate the share of rewards that have not been seen by the holder
-    // rewards account.
-    let rewards_to_harvest = {
-        // Rewards accrued by the system that haven't been seen by the holder.
-        let unseen_total_rewards = current_total_rewards
-            .checked_sub(last_seen_total_rewards)
-            .ok_or(ProgramError::ArithmeticOverflow)?;
-
-        // The holder's share of the unseen rewards.
-        let unseen_unharvested_rewards =
-            calculate_reward_share(token_supply, token_account_balance, unseen_total_rewards)?;
-
-        // The total unharvested rewards the holder is eligible to claim.
-        let total_eligible_rewards = holder_rewards_state
-            .unharvested_rewards
-            .checked_add(unseen_unharvested_rewards)
-            .ok_or(ProgramError::ArithmeticOverflow)?;
-
-        // Temporarily update unharvested rewards to reflect total eligible
-        // rewards. This will be re-adjusted post-harvest.
-        holder_rewards_state.unharvested_rewards = total_eligible_rewards;
-
-        // Make sure the pool can't be overdrawn by harvesting.
+    // Determine the amount the holder can harvest.
+    //
+    // This is done by subtracting the `last_rewards_per_token` rate from the
+    // pool's current rate, then multiplying by the token account balance.
+    //
+    // If the pool doesn't have enough lamports to cover the rewards, only
+    // harvest the available lamports. This should never happen, but the check
+    // is a failsafe.
+    let pool_excess_lamports = {
         let rent = <Rent as Sysvar>::get()?;
         let rent_exempt_lamports = rent.minimum_balance(std::mem::size_of::<HolderRewardsPool>());
-
-        // If the pool doesn't have enough lamports to cover the rewards, only
-        // harvest the available lamports.
-        total_eligible_rewards.min(
-            holder_rewards_pool_info
-                .lamports()
-                .checked_sub(rent_exempt_lamports)
-                .ok_or(ProgramError::ArithmeticOverflow)?,
-        )
+        holder_rewards_pool_info
+            .lamports()
+            .saturating_sub(rent_exempt_lamports)
     };
+    let eligible_rewards = calculate_eligible_rewards(
+        pool_state.rewards_per_token,
+        holder_rewards_state.last_rewards_per_token,
+        token_account_balance,
+    )?;
+    let rewards_to_harvest = eligible_rewards.min(pool_excess_lamports);
```

Comment:
Since our math is good now, as a courtesy to users, how about not updating if `rewards_to_harvest` is 0? That way small holders never lose out on rewards.

I'm thinking we can also add a flag to this instruction to optionally return an error. That way, scripts that harvest rewards for users can pack big transactions to update harvest for loads of users at once, and won't abort if one of those has no rewards. But for other situations, people can get an error if they want.

What do you think?

Reply:
Sorry I think I'm missing how anyone would miss out on rewards here. If they're entitled to zero because that's just what the calculation is based on their last seen rate, it's a no-op. If they're eligible for zero because the pool is underfunded, it updates their unharvested rewards with the rest that they couldn't harvest. Maybe that's an issue in and of itself.

Regardless, in general I think returning a `NoRewardsToHarvest` error when that value is zero is okay.

Reply:
Oh, maybe this should actually be:

```rust
holder_rewards_state.unharvested_rewards = unharvested_rewards - rewards_to_harvest;
```

Reply:
Not updating the last seen is the whole point. Imagine there are 10 tokens, and I have 1, and I've just harvested. If 1 lamport is distributed, then I'm entitled to 1/10 of that. But we can't subdivide lamports! So if I call harvest again, and the last seen gets updated, I lose out on 1/10 of a lamport.

If instead, nothing happens, then I can keep calling it with no effect, but once a total of 10 lamports have been distributed, then I'll get the 1 lamport in rewards that I'm entitled to.

Reply:
Okay great, makes sense! I'll add this.